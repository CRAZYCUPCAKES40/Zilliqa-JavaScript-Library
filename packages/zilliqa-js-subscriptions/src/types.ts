//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program.  If not, see <https://www.gnu.org/licenses/>.

export type QueryParam = 'NewBlock' | 'EventLog';

export const enum SocketConnect {
  READY = 'ready',
  CONNECT = 'connect',
  ERROR = 'error',
  CLOSE = 'close',
}

export const enum SocketState {
  SOCKET_CONNECT = 'socket_connect',
  SOCKET_MESSAGE = 'socket_message',
  SOCKET_READY = 'socket_ready',
  SOCKET_CLOSE = 'socket_close',
  SOCKET_ERROR = 'socket_error',
}

export const enum EventType {
  NEW_BLOCK = 'NewBlock',
  EVENT_LOG = 'EventLog',
  NOTIFICATION = 'notification',
  UNSUBSCRIBE = 'Unsubscribe',
}

export interface NewBlockQuery {
  query: string;
}

export interface NewEventQuery {
  query: string;
  addresses: string[];
}

export interface UnsubscribeNewBlock {
  query: 'Unsubscribe';
  type: 'NewBlock';
}

export interface UnsubscribeEventLog {
  query: 'Unsubscribe';
  type: 'EventLog';
}
