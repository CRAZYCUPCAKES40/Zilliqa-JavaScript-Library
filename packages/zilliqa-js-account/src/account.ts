import * as zcrypto from '@zilliqa-js/crypto';

export class Account {
  /**
   * fromFile
   *
   * Takes a JSON-encoded keystore and passphrase, returning a fully
   * instantiated Account instance.
   *
   * @param {string} file
   * @param {string} passphrase
   * @returns {Promise<Account>}
   */
  static async fromFile(file: string, passphrase: string): Promise<Account> {
    try {
      const keystore = JSON.parse(file);
      const privateKey = await zcrypto.decryptPrivateKey(passphrase, keystore);

      return new Account(privateKey);
    } catch (err) {
      throw new Error(`Could not decrypt keystore file.`);
    }
  }

  privateKey: string;
  publicKey: string;
  address: string;

  get checksumAddress(): string {
    try {
      return zcrypto.getAddress(this.address).checkSum;
    } catch (error) {
      throw error;
    }
  }

  get bech32Address(): string {
    try {
      return zcrypto.getAddress(this.address).bech32;
    } catch (error) {
      throw error;
    }
  }
  get bech32TestNetAddress(): string {
    try {
      return zcrypto.getAddress(this.address).bech32TestNet;
    } catch (error) {
      throw error;
    }
  }
  get base58Address(): string {
    try {
      return zcrypto.getAddress(this.address).base58;
    } catch (error) {
      throw error;
    }
  }
  get hexAddress(): string {
    try {
      return zcrypto.getAddress(this.address).bytes20Hex;
    } catch (error) {
      throw error;
    }
  }
  constructor(privateKey: string) {
    this.privateKey = privateKey;
    this.publicKey = zcrypto.getPubKeyFromPrivateKey(this.privateKey);
    this.address = zcrypto.getAddressFromPublicKey(this.publicKey);
  }

  /**
   * toFile
   *
   * @param {string} passphrase
   * @param {kdf} 'pbkdf2' | 'scrypt'
   * @returns {Promise<string>}
   */
  async toFile(
    passphrase: string,
    kdf: 'pbkdf2' | 'scrypt' = 'scrypt',
  ): Promise<string> {
    if (!passphrase || !passphrase.length) {
      throw new Error('Passphrase cannot have a length of 0');
    }

    const keystore = await zcrypto.encryptPrivateKey(
      kdf,
      this.privateKey,
      passphrase,
    );

    return keystore;
  }

  /**
   * signTransaction
   *
   * @param {Buffer} bytes - the data to be signed
   *
   * @returns {string} - the hex encoded signature. it is a concatenation of
   * the r and s values in hex, each padded to a length of 64.
   */
  signTransaction(bytes: Buffer) {
    return zcrypto.sign(bytes, this.privateKey, this.publicKey);
  }
}
