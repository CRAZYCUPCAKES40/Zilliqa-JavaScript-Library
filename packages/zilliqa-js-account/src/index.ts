import * as util from './util';

export * from './account';
export * from './wallet';
export * from './transaction';
export * from './types';

export { util };
