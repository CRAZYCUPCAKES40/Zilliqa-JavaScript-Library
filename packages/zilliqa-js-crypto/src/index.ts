import elliptic from 'elliptic';

import { randomBytes } from './random';
import * as schnorr from './schnorr';

const HEX_ENC: 'hex' = 'hex';
const secp256k1 = elliptic.ec('secp256k1');

/**
 * generatePrivateKey
 *
 * @returns {string} - the hex-encoded private key
 */
export const generatePrivateKey = (): string => {
  return secp256k1
    .genKeyPair({
      entropy: randomBytes(secp256k1.curve.n.byteLength()),
      entropyEnc: HEX_ENC,
      pers: 'zilliqajs+secp256k1+SHA256',
    })
    .getPrivate(HEX_ENC);
};

/**
 * sign
 *
 * @param {string} hash - hex-encoded hash of the data to be signed
 *
 * @returns {string} the signature
 */
export const sign = (
  msg: Buffer,
  privateKey: string,
  pubKey: string,
): string => {
  const sig = schnorr.sign(
    msg,
    Buffer.from(privateKey, 'hex'),
    Buffer.from(pubKey, 'hex'),
  );

  let r = sig.r.toString('hex');
  let s = sig.s.toString('hex');
  while (r.length < 64) {
    r = '0' + r;
  }
  while (s.length < 64) {
    s = '0' + s;
  }

  return r + s;
};

export { schnorr };
export * from './util';
export * from './keystore';
export * from './random';
export * from './types';
export * from './signature';
