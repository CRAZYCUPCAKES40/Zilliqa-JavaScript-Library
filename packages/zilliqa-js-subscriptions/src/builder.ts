import { NewTxBlockSubscription } from './newblock';
import { NewEventSubscription } from './newevent';
import { SubscriptionOption } from './types';

export class SubscriptionBuilder {
  buildNewBlockSubscriptions(url: string, options?: SubscriptionOption) {
    return new NewTxBlockSubscription(url, options);
  }

  buildEventLogSubscriptions(url: string, options?: SubscriptionOption) {
    return new NewEventSubscription(url, options);
  }
}
