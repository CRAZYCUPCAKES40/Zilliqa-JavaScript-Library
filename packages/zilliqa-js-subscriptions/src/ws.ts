//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program.  If not, see <https://www.gnu.org/licenses/>.

import mitt from 'mitt';
import { w3cwebsocket as W3CWebsocket } from 'websocket';
import {
  NewBlockQuery,
  NewEventQuery,
  SocketConnect,
  SocketState,
} from './types';

export class WebSocketProvider {
  public static NewWebSocket(
    url: string,
    options: any = {},
  ): WebSocket | W3CWebsocket {
    // tslint:disable-next-line: no-string-literal
    if (typeof window !== 'undefined' && (<any>window).WebSocket) {
      // tslint:disable-next-line: no-string-literal
      return new WebSocket(url, options.protocol);
    } else {
      const headers = options.headers || {};
      const urlObject = new URL(url);

      if (!headers.authorization && urlObject.username && urlObject.password) {
        const authToken = Buffer.from(
          `${urlObject.username}:${urlObject.password}`,
        ).toString('base64');
        headers.authorization = `Basic ${authToken}`;
      }

      return new W3CWebsocket(
        url,
        options.protocol,
        undefined,
        headers,
        undefined,
        options.clientConfig,
      );
    }
  }

  url: string;
  emitter: mitt.Emitter;
  handlers: any = {};
  websocket: WebSocket | W3CWebsocket;
  // use concrete object instead of `any`
  subscriptions: any;

  // basically, options is a collection of metadata things like protocol or headers
  constructor(url: string, options: any = {}) {
    this.url = url;
    this.emitter = mitt(this.handlers);
    this.websocket = WebSocketProvider.NewWebSocket(url, options);
    this.subscriptions = {};
    // todo register event listener
    this.websocket.onopen = this.onConnect.bind(this);
    this.websocket.onclose = this.onClose.bind(this);
    this.websocket.onmessage = this.onMessage.bind(this);
  }

  onClose(closeEvent: any) {
    // todo impl this
    console.log('close');
    this.websocket.close();
    return;
  }

  async onConnect() {
    if (!this.subscriptions) {
      this.subscriptions = {};
    }
    // todo handle reconnect issue
    this.emitter.emit(SocketState.SOCKET_CONNECT);
    this.emitter.emit(SocketConnect.CONNECT);
  }

  onMessage(msg: MessageEvent) {
    // todo impl this
    console.log(msg);
  }

  addEventListener(type: string, handler: mitt.Handler) {
    this.emitter.on(type, handler);
  }

  removeEventListener(type: string, handler: mitt.Handler) {
    this.emitter.off(type, handler);
  }

  connecting() {
    return this.websocket.readyState === this.websocket.CONNECTING;
  }

  send(query: NewBlockQuery | NewEventQuery): Promise<any> {
    return new Promise((resolve, reject) => {
      if (!this.connecting()) {
        try {
          this.websocket.send(JSON.stringify(query));
        } catch (error) {
          throw error;
        }
        this.emitter.on(query.query, (data) => {
          resolve(data);
        });
        this.emitter.on(SocketConnect.ERROR, reject);
      }
      // may be emitted inside onConnect
      this.emitter.on(SocketConnect.CONNECT, () => {
        this.send(query)
          .then(resolve)
          .catch(reject);
      });
    });
  }

  async subscribe(payload: NewBlockQuery | NewEventQuery) {
    const response = await this.send(payload);
    // todo tell if this subscribe action caused correct result
    this.subscriptions[response.result] = {
      id: response.result,
      parameters: payload.query,
    };
    return response.result;
  }
}
