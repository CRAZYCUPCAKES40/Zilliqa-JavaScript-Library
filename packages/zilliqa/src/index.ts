import { HTTPProvider, Provider } from '@zilliqa-js/core';
import { TransactionFactory, Wallet } from '@zilliqa-js/account';
import { Contracts } from '@zilliqa-js/contract';
import { Blockchain, Network } from '@zilliqa-js/blockchain';
import { HRP, tHRP } from '@zilliqa-js/crypto';
import { ZilConfig, ChainType, ChainID } from './utils';

export class Zilliqa {
  provider: Provider;

  blockchain: Blockchain;
  network: Network;
  contracts: Contracts;
  transactions: TransactionFactory;
  wallet: Wallet;

  constructor(node: string, provider?: Provider, config?: ZilConfig) {
    let zilProvider;

    if (config) {
      const zilConfig = this.getConfig(config);
      zilProvider = zilConfig.provider;
    }

    this.provider =
      provider || zilProvider || new HTTPProvider(node, ChainID.MainNet);
    this.wallet = new Wallet(this.provider);
    this.blockchain = new Blockchain(this.provider, this.wallet);
    this.network = new Network(this.provider, this.wallet);
    this.contracts = new Contracts(this.provider, this.wallet);
    this.transactions = new TransactionFactory(this.provider, this.wallet);
  }

  getConfig(config: ZilConfig) {
    const { chainType, endpoint } = config;
    let chainID = ChainID.MainNet;
    let hrp = HRP;

    switch (chainType) {
      case ChainType.MainNet: {
        chainID = ChainID.MainNet;
        hrp = HRP;
        break;
      }
      case ChainType.TestNet: {
        chainID = ChainID.TestNet;
        hrp = tHRP;
        break;
      }
      default:
        break;
    }
    const provider = new HTTPProvider(endpoint, chainID);
    return { provider, chainID, hrp };
  }
}
