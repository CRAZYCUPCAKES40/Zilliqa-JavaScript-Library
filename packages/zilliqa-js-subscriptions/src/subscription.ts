import { WebSocketProvider } from './ws';

export class Subscription extends WebSocketProvider {
  async subscribe() {
    // todo
    return this;
  }
}
