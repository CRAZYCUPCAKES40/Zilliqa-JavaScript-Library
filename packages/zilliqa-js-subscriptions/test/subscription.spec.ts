import { Server } from 'mock-socket';
import { WebSocketProvider } from '../src';
// import {WebSocketProvider} from '../dist/ws';

describe('WebSocketProvider', () => {
  const fakeURL = 'ws://localhost:8080';

  it('should be able to connect to websocket server', async () => {
    const mockServer = new Server(fakeURL);
    mockServer.on('connection', (socket) => {
      socket.on('message', (data) => {
        console.log('data = ', data);
        expect(data).toEqual('{"query":"1000"}');
        socket.send('test message from mock server');
        socket.close();
      });
    });

    const ws = new WebSocketProvider(fakeURL);
    expect(ws.subscribe({ query: '1000' })).resolves.toReturn();
    // following code will cause timeout error which makes this unit test failed
    // I have not reach out a good way to let this unit test succeed, but indeed it is a good tool test our class for now
    // await expect(ws.subscribe({query: '1000'})).resolves.toReturn();
  });
});
