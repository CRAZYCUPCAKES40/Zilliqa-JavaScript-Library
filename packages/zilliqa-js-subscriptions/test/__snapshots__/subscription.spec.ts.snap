// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebSocketProvider should be able to connect to websocket server 1`] = `"received is not a function"`;
